### PR TITLE
Dump additional apis

### PR DESCRIPTION
**Pull request by afranke**
Currently, only the client-server is made available as an OpenAPI JSON file. This adds support for the rest of the API.

* `application-service` and `push-gateway` work.
* `server-server` does not work yet. It fails to resolve a specific reference, still investigating.
* `identity` does not work yet. I haven’t had time to look into it yet. YAML reading works fine, the problem occurs when writing the JSON output.

Work towards https://github.com/matrix-org/matrix-doc/issues/1458 ~ TravisR

Preview: https://pr3684--matrix-org-previews.netlify.app

### CODE REVIEW

**Comment by afranke** [comment]
@turt2live I’d like this to be included with the v1.2 release, so the new JSON is made available (and so that Rapidoc can pick them up and allow switching between them). I don’t know if I’ll be able to fix the `server-server` and `identity` generation by then, but the PR could be merged as it is (and the missing API could be added later on). 2 new API JSON files is better than none.

**Comment by turt2live** [comment]
It looks like this is a draft. Is it ready for review?

**Comment by afranke** [comment]
I set it as draft because of the missing API, but we can review and merge if that works for you without them.

File: scripts/dump-swagger.py [hunk 1]
```diff
@@ -87,11 +87,24 @@ def edit_links(node, base_url):
...
+    help="""The spec release version to generate for. Default:
     %(default)s""",
 )
+available_apis = {
+        "client-server": "Matrix Client-Server API",
+        "server-server": "Matrix Server-Server API",
+        "application-service": "Matrix Application Service API",
+        "identity": "Matrix Identity Service API",
+        "push-gateway": "Matrix Push Gateway API",
+        }
```
Comment (turt2live) [inline scripts/dump-swagger.py:101]: nitpick: indentation looks off, but also I'm not a python lint expert (we should also probably actually have a linter, but that's a future problem).

**Review by turt2live — COMMENTED** [review]
Overall this looks fine, and is a substantial improvement towards having a viewer for all the APIs.

The server-server definitions are significantly more complex than the client-server definitions, so I'm not entirely surprised that it has issues. Identity should be closer to appservices though, so I'm a bit surprised on that. We might have a slightly invalid bit of yaml somewhere that causes invalid json on the output (yaml parsers are a bit less strict than json).

At a glance the outputted JSON looks reasonable as well.

I'd be comfortable landing this as-is in v1.2, though if we can get identity and/or s2s working before Wednesday my-morning then all the better. Will merge then assuming changes look about the same/okay ;)

File: scripts/dump-swagger.py (hunk 1 above)
Comment (afranke) [reply scripts/dump-swagger.py]: Agreed. I usually apply `black` everywhere on my projects, but it can be a bit noisy on existing ones. Fixed.

**Comment by afranke** [comment]
The `server-server` issue comes from #3689.

**Comment by afranke** [comment]
So… it looks like the server-server is not valid and we didn’t know about it because CI didn’t try to validate it so far.

We have places with `"properties": []` when it should be `"properties": {}`. I can fix those, and it looks like that wasn’t valid JSON Schema to begin with. Do I need to add a spec changelog for that? Do I have to check anything else with regards to synapse or such things?

**Comment by richvdh** [comment]
I think if there are no user-visible changes, you don't need a changelog.

**Comment by afranke** [comment]
At this point, I’m thinking a JSON file that doesn’t pass validation is better than no file at all, especially considering that the errors have been there for a while and we just didn’t know about them. Don’t you think?

We can spend time fixing the validation errors, but I am not sure we will get all of them resolved by the time we release v1.2. I propose we remove the server-server validation for now to make CI happy, and track the server-server validation errors in a new issue & PR. What do you reckon?

**Comment by turt2live** [comment]
Generating data that is invalid and also isn't covered by CI is far from ideal - if we can make it valid, great, but if not then let's just not generate the server-server json at all until it can be fixed.

**Comment by afranke** [comment]
Just to make it clear: the CI says the server-server **YAML** file is invalid. That is *slightly* different from “generating data that is invalvid”. That also means that the data consumed by hugo (and potentially 3rd parties) is already invalid.

**Comment by turt2live** [comment]
Looking at the error from CI (and reproduced locally), I don't think it's invalid data. It looks like the `$ref` handling is relative to the wrong directory still.

**Comment by afranke** [comment]
That seems to be https://github.com/APIDevTools/json-schema-ref-parser/issues/251

**Comment by afranke** [comment]
I ran the validator locally on all server-server files and, aside from the path issue, 12 out of 22 are not valid (with genuine errors, not `$ref` handling). There is no way I’m going to fix this quick, so I am removing the server-server API for now, and it can be added later on. With this change, I am removing the draft status as this is ready to be merged.

**Review by turt2live — APPROVED** [review]
thanks! Here's hoping the s2s stuff isn't too painful 😬

**Pull request merged**